packages/grafana-prometheus/src/querycache/QueryCache.ts

```diff
@@ -216,11 +218,25 @@ export class QueryCache<T extends SupportedQueryTypes> {
         // trim all frames to in-view range, evict those that end up with 0 length
         let nonEmptyCachedFrames: DataFrame[] = [];
 
+        // series within a query
         cachedFrames.forEach((frame) => {
           // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
           let table: Table = frame.fields.map((field) => field.values) as Table;
 
-          let trimmed = trimTable(table, newFrom, newTo);
+          // Warning! logic specific to prometheus.
+          const target = request.targets.find((t) => t.refId === respFrames[0].refId);
+
+          // @todo needs unit testing
+          let dataPointStep = request.intervalMs;
+          if (target?.interval) {
+            const minStepMs = rangeUtil.intervalToMs(target.interval);
+            if (minStepMs > request.intervalMs) {
+              dataPointStep = minStepMs;
+            }
+          }
```

Comment:
if this is prom-specific, maybe it can be extracted into a function that returns the `fromOffset` or something (default is 0), and below can be.

```ts
let trimmed = trimTable(table, newFrom - fromOffset, newTo);
```